### ISSUE TITLE
Jhipster:Clevercloud sub generator

### ISSUE THREAD

**Issue by savoisn** [jhipster/generator-jhipster#1655]
I've been working a lot with jhipster docker and clevercloud recently.
Would you guys like me to try to create a PR to work on that?

**Comment by phouverneyuff** [comment]
The pricing isn't good.

https://www.clever-cloud.com/en/pricing/

**Comment by jdubois** [comment]
- Docker would be good, but there are many ways to do it -> so yes I'm interested, but it depends on how you implement it, of course!
- CleverCloud is interesting, but as it's a paying service, I can't test it (same issue I have with Oracle, for example, and that's really annoying for me). Besides, I don't want to spend time maintaining stuff when it's a company making money on it (please note I'm doing all this for free, this is not my day job). So we could integrate it with some warning, like "this is maintained by CleverCloud and not JHipster", but then I would need some commitment from CleverCloud (I don't want to have a broken sub-generator around, and I need someone to assign the issues to). I guess you work at CleverCloud @savoisn ?

**Comment by savoisn** [comment]
Hum no not working for them at all... but I wanted to test things with docker and they are nearly the only one providing it... (of course there's google but it's very complicated to set up), but I understand your point competently. 

Concerning docker 

 - as you say there's a lot of way to implement it. I've been playing with some of those solutions and I am sure there's a lot I've haven't even seen.
 - it also depends on the way you want to deploy your docker, 
	- is it just a dockerfile and the saas is constructing all the layers (clever cloud approach), which does not allow to package the produced war
	- are we constructing the images locally and pushing it to the saas (google approach, very bad in terms of deployment time imo - 7 hours on my first test...)

I had in mind to create a Dockerfile from which you clone a git repo with the jhipster project, and run the npm install and mvn package -Pprod to produce the war but this require a really big images like the one you already provided us. 
And maybe even do a docker build after that to create the images.

But everything else might very dependent of the docker platform we will deploy the application

What I've came out with is that :
 - produce war inside your docker image (jdubois/jhipster-docker)
 - push war to Bintray, i like to keep history of my build (might be a good subgenerator also)
 - created a Dockerfile from ubuntu, install java, wget war from Bintray and run java -jar
 - produced a relatively small docker image (800Mo), relatively fast build (2 minutes or something)
we good replace bintray with a nexus...

I need time to test with the official docker-java image but as it provieds an openJDK I didn't want to have other issues interfering. (it should be ok because heroku is using it but i've been shy on that hehe)

As you can see I am not very fixed on the way I want to deploy with docker, and so I am really open to any proposition.

**Comment by waxzce** [comment]
Hi @savoisn & @jdubois 

First Thank a lot to @savoisn to bring some jhipster goodness to Clever Cloud. It's really great :-) I'm really happy when the community is helping Clever Cloud to be more efficient for everyone?

Jhipster is there to help people creating project quickly with java, and Clever Cloud is here to help people creating a service quickly without managing any hosting-related issues, and issue is quite a problem with java based solutions.

Sure, Clever Cloud is not free, and there is [very good reasons for this](https://www.clever-cloud.com/blog/company/2015/04/14/why-isnt-t-there-a-free-tier-in-clever-cloud/) (BTW others PaaS following us, and heroku is not free anymore). We have a free trial of 20€ for every new subscribers, we regularly extend it when people ask the support to do it. A new offer (active on some runner already) will bring Clever Cloud for everyone with a less than 5$/month offer, pretty cheap isn't it?

So basically, I am happy that the community is now pushing some documentation to the frameworks, like Play! for example. We can be vigilant on this and make regular tests. BTW we will hire more and more engineers committing exclusively on open source devs tools starting next month.

FYI: to support open source we provide Clever Cloud hosting for free to open source project applying for it, and it will be a pleasure to support jhipster in this way :-)

**Comment by jdubois** [comment]
@savoisn OK the Docker build should be discussed -> we should separate this from the clevercloud sub-generator.
My proposal is that you open a specific ticket for Docker, and ask for people's opinions. You can even do a chat with all people interested in the [chat room](https://gitter.im/jhipster/generator-jhipster) -> if you need it I can tweet the date/hour of the talk, to attract more people

For CleverCloud -> @waxzce if you support the sub-generator yourself, I'll happily merge it. Otherwise I won't have the time to support it, as we already have 3 different cloud providers to support and that's a lot of work already.

**Comment by phouverneyuff** [comment]
I had already created a script to run docker with jhipster and mysql.
This script download, configure, start, stop and enter in shell.

I can give the url of my github for you take a look.

**Comment by waxzce** [comment]
ATM  we let @savoisn be lead and he will fwd us the lead if he left. Works for all ?

**Comment by jdubois** [comment]
@phouverneyuff can you participate in #1674 for the Docker build? For me this is the most important, as many more people use Docker.

For CleverCloud -> @savoisn I don't think you & me can work on both Docker and Clevercloud at the same time (at least, I won't have the time), so can we make the Docker build work first? And then add CleverCloud?

**Comment by jdubois** [comment]
This has been opened for more than 1 month, is anybody working on this? Otherwise I will close it.

**Comment by jdubois** [comment]
This has been opened for 2 months without work, so I'm closing this.

**Comment by ldoguin** [comment]
I would be happy to work on this and support it, can we reopen it @jdubois ?

**Comment by jdubois** [comment]
@ldoguin just to have a clean ticket, could you create a new one? This is so old, so many things have changed... And then of course no problem

**Comment by deepu105** [comment]
I would suggest to start this as Module instaed of a PR here


Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
Oh yes @deepu105 is right -> can you start as a module? This will allow you to go faster (without us, as we have way too much work at the moment to do code reviews and merges), and when it works fine, it's easy to integrate in the main generator (just a copy/paste)

**Comment by ldoguin** [comment]
Thanks guys, I am on it :)

**Comment by waxzce** [comment]
:-)
